File: 2주차/SWEA1868. 파핑파핑 지뢰찾기/최원재.py
T = int(input())
 
dx = [-1, -1, -1, 0, 0, 1, 1, 1]
dy = [-1, 0, 1, -1, 1, -1, 0, 1]
 
def cnt_mine(r, c):
    cnt = 0; dr = 0; dc = 0;
    for i in range(8):
        dr = r + dx[i]
        dc = c + dy[i]
        if dr < 0 or dr >= n or dc < 0 or dc >= n:
            continue
        if mine_map[dr][dc] == '*':
            cnt+=1;
    return cnt;
 
def open_mine(r, c):
    cnt = 0; dr = 0; dc = 0;
    num_map[r][c] = -1
    for i in range(8):
        dr = r + dx[i]
        dc = c + dy[i]
        if dr < 0 or dr >= n or dc < 0 or dc >= n:
            continue
        if num_map[dr][dc] == 0:
            open_mine(dr, dc)
        num_map[dr][dc] = -1
 
for test_case in range(1, T + 1):
    n = int(input())
    mine_map = [' ' for j in range(n)]
    num_map = [[0 for j in range(n)] for i in range(n)]
    for i in range(n):
        mine_map[i] = str(input())
    for r in range(n):
        for c in range(n):
            if mine_map[r][c] == '*':
                num_map[r][c] = -1
            else:
                num_map[r][c] = cnt_mine(r, c)
    cnt = 0
    for r in range(n):
        for c in range(n):
            if num_map[r][c] == 0:
                cnt+=1
                open_mine(r, c)
 
    for r in range(n):
        for c in range(n):
            if num_map[r][c] != -1:
                cnt+=1
    print("#", test_case," ",cnt,sep="")

File: 2주차/SWEA1868. 파핑파핑 지뢰찾기/최원재.java
import java.util.Scanner;
 
class Solution
{
     
    static char[][] mine_map;
    static int[][] num_map;
     
    static int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};
    static int[] dy = {-1, 0, 1, -1, 1, -1, 0, 1};
     
    static int n;
     
    static int cnt_mine(int r, int c) {
        int cnt = 0, dr = 0, dc = 0;
        for (int i = 0; i<8; i++) {
            dr = r + dx[i];
            dc = c + dy[i];
            if (dr < 0 || dr >= n || dc < 0 || dc >= n) continue;
             
            if (mine_map[dr][dc] == '*') cnt+=1;
        }
        return cnt;
    }
     
    static void open(int r, int c) {
        int cnt = 0, dr = 0, dc = 0;
        num_map[r][c] = -1;
        for (int i = 0; i < 8; i++) {
            dr = r + dx[i];
            dc = c + dy[i];
            if (dr < 0 || dr >= n || dc < 0 || dc >= n) continue;
            if (num_map[dr][dc] == 0) open(dr, dc);
            num_map[dr][dc] = -1;
        }
    }
     
    public static void main(String args[]) throws Exception
    {
        Scanner sc = new Scanner(System.in);
         
        int T = sc.nextInt();
         
        for(int test_case = 1; test_case <= T; test_case++)
        {
            n = sc.nextInt();
            String s;
            mine_map = new char[n][n];
            num_map = new int[n][n];
            for(int i = 0; i < n; i++) {
                mine_map[i] =  sc.next().toCharArray();
                 
            }
            for(int i = 0; i<n;i++) {
                for(int j = 0 ;j<n;j++) {
                    if(mine_map[i][j] == '*') num_map[i][j] = -1;
                    else num_map[i][j] = cnt_mine(i, j);
                }
            }
            int cnt = 0;
            for(int i = 0; i<n;i++) {
                for(int j = 0 ;j<n;j++) {
                    if(num_map[i][j] == 0) {
                        cnt++;
                        open(i,j);
                    }
                }
            }
            for(int i = 0; i<n;i++) {
                for(int j = 0 ;j<n;j++) {
                    if(num_map[i][j] != -1) {
                        cnt++;
                    }
                }
            }       
            System.out.println("#" + test_case + " " + cnt);
        }
         
    }
     
}
